### PR TITLE
Support nn.RNN in torch.jit.script (#28418)

### PR DESCRIPTION
**Pull request by you74674**
I copied and modified all methods from GRU to RNN in order to support nn.RNN in torch.jit.script.  (#28418)
There are three parts that are different from the original implementation in RNNBase, but basically all of them are because of the same reason as GRU and LSTM.
1. Deal with PackedSequence type check in script.
2. Inline \_VF.rnn\_(tanh|relu).
3. Add overload of forward.

### CODE REVIEW

File: torch/nn/modules/rnn.py [hunk 1]
```diff
@@ -396,6 +398,65 @@ def __init__(self, *args, **kwargs):
...
+            hx = torch.zeros(self.num_layers * num_directions,
+                             max_batch_size, self.hidden_size,
+                             dtype=input.dtype, device=input.device)
+        else:
+            # Each batch of the hidden state should match the input sequence that
+            # the user believes he/she is passing in.
+            hx = self.permute_hidden(hx, sorted_indices)
+
+        self.check_forward_args(input, hx, batch_sizes)
+        if self.mode == "RNN_TANH":
```
Comment (eellison) [inline torch/nn/modules/rnn.py:437]: Can you initialize `_VF.rnn_tanh` / `_VF.rnn_relu` in the initialization as `self.activation` ? And then do the same with `nn.GRU`? I think if you did that you could have `RNN` and `GRU` extend from the same base class that defines the forward method, and then just separately define `self.activation` in the `__init__`

It's possible that `_overload_method` doesn't work yet with inheritance and that doesn't work though...

**Review by eellison — COMMENTED** [review]
Hi @you74674, thanks for the PR! 

Do you mind adding some tests for this ? you can add to `test_export_rnn` in `test_jit.py`.

I have a couple comments to better share code.

**Comment by you74674** [comment]
In 1.4.0 it seems like that it's not allowed to have this kind of function as an attribute of jit module, but I'm not sure about the latest version. (Or maybe I just miss something needed for it to pass?)

About the test, I have tested that the new implemented forward gives the same result as the old forward in RNNBase, with both relu and tanh, so I think I can add it. Is it fine to just add this test either before or after the for loop of `test_export_rnn` function? Since it is only about RNN, but has nothing to do with GRU.
And, is there anything else I should test?

File: torch/nn/modules/rnn.py (hunk 1 above)
Comment (you74674) [reply torch/nn/modules/rnn.py]: I tried the overload with inheritance and it looks like that doesn't work.
https://gist.github.com/you74674/48dd48ba2b21cc5af40aa512a43b4f32

**Review by eellison — APPROVED** [review]
LGTM. It's unfortunate that we have to copy the forward, but I'm going to try to get the overloads with inheritance working soon to avoid that. I'm not sure if it's worth holding up this PR in the meantime...

**Comment by you74674** [comment]
Thanks. Let me know if there's something I still need to do about this PR. (Although it looks like I have already messed it up a little... sorry about that, I'm not familiar with PR yet.)

**Comment by eellison** [comment]
@you74674 nope i don't think you messed it up! just existing holes in our infrastructure

**Pull request closed without merging**